[Lee Chun Wei] iP

chunweii opened the pull request:
# Duke

> The key is not to prioritize what’s on your schedule, but to schedule your priorities.  
> ~ Stephen Covey

Duke is an awesome way to keep track of all your daily tasks.

Forgot what is coming up in your schedule? **Duke it**
Hard to remember all your appointments? **Duke it**
Find it difficult to manage your todo list? **Duke it**

Duke is:
1. `Text-based`
2. Easy-to-learn
3. _Blazing_ fast :fire:
4. **Free** :no_entry_sign::dollar:

## Setting up

Prerequisites: JDK 11, download [here](https://www.oracle.com/java/technologies/javase-jdk11-downloads.html).

1. Download the Duke jar file [here](https://github.com/chunweii/ip/releases)
2. Open a terminal/command prompt and navigate to the directory of the jar file.
3. Run `java -jar ip.jar`
4. Add your tasks!
## Future features
- ~~Graphical User Interface~~ will be implemented in the next release!
- Reminders

Do support us! If you are a Java programmer, consider improving the program and submitting a pull request! Here is our main method:
```java
public static void main(String[] args) {
    // new Duke(<file path>, <limit>).run();
    new Duke("./dukedata.txt", 100).run();
}
```

Commit:
Add Gradle support

chunweii committed:
Add basic skeletal chatbot

chunweii committed:
Add list functionality

chunweii committed:
Add checklist functionality

chunweii committed:
Add todo, deadline and event

chunweii committed:
Add leading space for list number

chunweii committed:
Add test case

chunweii committed:
Add missing description error and DukeException class

chunweii committed:
Add deletion capability

chunweii committed:
Abstract processInput and remove emoji from error message

chunweii committed:
Add comments and javadoc

chunweii committed:
Add save data functionality and abstract task update

chunweii committed:
Add date and time logic

Further, group the logic into packages

chunweii committed:
Add upcoming tasks logic and command

chunweii committed:
Abstract printing tasks logic

chunweii committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/duke/Duke.java

chunweii committed:
Clean up after merge

chunweii committed:
Add Command Parser, Storage, TaskEnums, TaskList

chunweii committed:
Clean up Duke.java

Remove all old and redundant methods/fields.

chunweii committed:
Fix text ui test bat

chunweii committed:
Add TaskList Junit test

chunweii committed:
Add Commands enums and Ui class

chunweii committed:
Remove large comments

chunweii committed:
Add jar file (currently in gitignore)

chunweii committed:
Clean up command enums

chunweii committed:
Add more javadoc

chunweii committed:
Edit javadoc

chunweii committed:
Edit code style to align with standard

chunweii committed:
Add find command

chunweii committed:
Merge branch 'branch-A-CodingStandard'

chunweii committed:
Fix merge issues

chunweii committed:
Fix bugs and rearrange code

Fix the issue with tasks not appearing and tasks not being printed with correct task number. Removed logic of getting upcoming tasks in the Ui class.

chunweii committed:
Change README.md

chunweii committed:
Merge remote-tracking branch 'origin/add-gradle-support' into week4

chunweii committed:
Add gradle

chunweii committed:
Add Checkstyle

chunweii committed:
Add basic GUI

chunweii committed:
Enhance GUI

chunweii committed:
Add FXML

chunweii committed:
Expand Storage class to enable loading of files

chunweii committed:
Enable resizing of GUI

chunweii committed:
Merge pull request #1 from chunweii/branch-Level-10

Enable resizing of GUI

sueann-chua commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * Prints all upcoming tasks from a list of tasks and task numbers.
+     *
+     * @param tasksWithTaskNumbers the list of tasks and task numbers, each in a SimpleImmutableEntry
```
Should end the sentence with a fullstop

sueann-chua commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+        return true;
+    }
+
+    private boolean validTaskNumber(int taskNumber) {
+        return taskNumber <= tasks.size() && taskNumber > 0;
+    }
```
boolean methods should be made to sound like booleans eg isValidTaskNumber

sueann-chua commented on src/main/java/duke/logic/LCommandsEnum.java:
```diff
@@ -0,0 +1,99 @@
+public enum LCommandsEnum {
+    BYE() {
+        @Override
+        public String helpMessage() {
+            return "Usage: bye\n" +
+                    "This will quit the Duke chatbot.";
```
Names representing methods must be verbs eg showHelpMessage

sueann-chua commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @return the size of the list as an integer.
+     */
+    public int size() {
+        return tasks.size();
+    }
```
names representing methods must be verbs

sueann-chua commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,173 @@
+    /**
+     * Gets all the tasks in the list containing the string in the description.
+     *
+     * @param pattern the pattern to be matched with the task descriptions in the list
+     * @return the immutable list of tasks containing str in the description, with their task numbers
```
Sentences should end with fullstops

sueann-chua reviewed (commented):
Overall, code is good and doesn't have many coding standard violations. Goodjob! :)

bingcheng45 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    public void run() {
+        Ui ui = new Ui();
+        while (!ui.willExit()) {
```
Maybe to add clarity, you can have ui.willContinue(), as having ! with will exit can be confusing at times.

bingcheng45 commented on src/main/java/duke/logic/LCommandParser.java:
```diff
@@ -0,0 +1,126 @@
+            if (inputArr.length < 2) {
+                throw new DukeException(TOO_LITTLE_ARGUMENTS_MESSAGE);
+            }
+            switch (commandEnum) {
+            case FIND:
+                ui.printTasksContaining(inputArr[1],
+                        taskList.getTasksContaining(inputArr[1]), taskList.size());
+                break;
+            case TODO: // fallthrough intended
+            case EVENT: // fallthrough intended
+            case DEADLINE:
+                ui.addTaskMessage(addTask(commandEnum.name(), inputArr[1], taskList), taskList.size());
+                break;
+            default:
+                int taskNumber;
+                try {
+                    taskNumber = Integer.parseInt(inputArr[1]);
+                } catch (NumberFormatException e) {
+                    throw new DukeException(INVALID_NUMBER_MESSAGE);
+                }
+                switch (commandEnum) {
+                case DONE:
+                    if (!taskList.markAsDone(taskNumber)) { // task already marked as done
+                        throw new DukeException("You have already marked this task (%s) as done",
+                                taskList.getTask(taskNumber));
+                    }
+                    ui.markAsDoneMessage(taskList.getTask(taskNumber));
+                    break;
+                case DELETE:
+                    ui.removeTaskMessage(taskList.removeTask(taskNumber), taskList.size());
+                    break;
+                default:
+                    throw new DukeException(INVALID_COMMAND);
+                }
```
Good mentioned of intended fallthrough, just curious if this can be done in a single switch statement rather than nested switch statement

bingcheng45 commented on src/main/java/duke/logic/LStorage.java:
```diff
@@ -0,0 +1,49 @@
+package duke.logic;
+
+import duke.task.Task;
+import duke.task.TaskList;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
```
Java imports should come before package imports

bingcheng45 reviewed (commented):
LGTM, Overall good job on following the conventions aside from the one mentioned by me and sueann-chua. Love your use of wildcards.

chunweii committed:
Add end date and time to event

chunweii committed:
Add assertions to Main Window UI, logic

chunweii committed:
Improve code quality for parsers

Some methods were too long. This violated the code
quality checks since methods should remain short.
Some parts of the command parser were abstracted to
another private method to make it readable.

chunweii committed:
Improve overall code quality for main

chunweii committed:
Use IntStream instead of for-loop in TaskList

chunweii committed:
Merge pull request #2 from chunweii/branch-A-Assertions

Add assertions to Main Window UI, logic

chunweii committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/task/Event.java

chunweii committed:
Merge branch 'master' into branch-A-Streams

chunweii committed:
Merge pull request #3 from chunweii/branch-A-Streams

Use streams in code

chunweii committed:
Merge branch 'master' into branch-A-CodeQuality

chunweii committed:
Merge pull request #4 from chunweii/branch-A-CodeQuality

Enhance code quality

chunweii committed:
Change System.exit to Platform.exit

chunweii committed:
Enhance help message

chunweii committed:
Edit README

chunweii committed:
Add many JUnit Tests

chunweii committed:
Add StorageStub to prevent file creation

chunweii committed:
Add Continuous Integration

chunweii committed:
Add Continuous Integration

chunweii committed:
Fix CI issues

chunweii committed:
Add User Guide

chunweii committed:
Fix circular image ratio and add exit delay

chunweii committed:
Change label style

chunweii committed:
Resolve truncation issue with labels

chunweii committed:
Refactor code

chunweii committed:
Edit user guide
